dpcpp/solver/batch_bicgstab_kernels.hpp.inc

```diff
@@ -0,0 +1,453 @@
+    }
+}
+
+
+template <const bool sg_kernel_all, typename ValueType>
+__dpct_inline__ void compute_alpha(const int num_rows, const ValueType& rho_new,
+                                   const ValueType* const r_hat_shared_entry,
+                                   const ValueType* const v_shared_entry,
+                                   ValueType& alpha, sycl::nd_item<3> item_ct1)
+{
+    auto sg = item_ct1.get_sub_group();
+    const auto sg_id = sg.get_group_id();
+    const auto tid = item_ct1.get_local_linear_id();
+    if constexpr (sg_kernel_all) {
+        if (sg_id == 0) {
+            single_rhs_compute_conj_dot_sg(num_rows, r_hat_shared_entry,
+                                           v_shared_entry, alpha, item_ct1);
+        }
+        item_ct1.barrier(sycl::access::fence_space::global_and_local);
+        if (tid == 0) {
+            alpha = rho_new / alpha;
+        }
+        item_ct1.barrier(sycl::access::fence_space::global_and_local);
+    } else {
+        single_rhs_compute_conj_dot(num_rows, r_hat_shared_entry,
+                                    v_shared_entry, alpha, item_ct1);
+        if (tid == 0) {
+            alpha = rho_new / alpha;
+        }
+    }
+}
+
+
+template <typename ValueType>
+__dpct_inline__ void update_s(const int num_rows,
+                              const ValueType* const r_shared_entry,
+                              const ValueType& alpha,
+                              const ValueType* const v_shared_entry,
+                              ValueType* const s_shared_entry,
+                              sycl::nd_item<3> item_ct1)
+{
+    for (int r = item_ct1.get_local_linear_id(); r < num_rows;
+         r += item_ct1.get_local_range().size()) {
+        s_shared_entry[r] = r_shared_entry[r] - alpha * v_shared_entry[r];
+    }
+}
+
+
+template <const bool sg_kernel_all, typename ValueType>
+__dpct_inline__ void compute_omega(const int num_rows,
+                                   const ValueType* const t_shared_entry,
+                                   const ValueType* const s_shared_entry,
+                                   ValueType& temp, ValueType& omega,
+                                   sycl::nd_item<3> item_ct1)
+{
+    auto sg = item_ct1.get_sub_group();
+    const auto sg_id = sg.get_group_id();
+    const auto tid = item_ct1.get_local_linear_id();
+    if constexpr (sg_kernel_all) {
+        if (sg_id == 0) {
+            single_rhs_compute_conj_dot_sg(num_rows, t_shared_entry,
+                                           s_shared_entry, omega, item_ct1);
+        } else if (sg_id == 1) {
+            single_rhs_compute_conj_dot_sg(num_rows, t_shared_entry,
+                                           t_shared_entry, temp, item_ct1);
+        }
+        item_ct1.barrier(sycl::access::fence_space::global_and_local);
+        if (tid == 0) {
+            omega /= temp;
+        }
+        item_ct1.barrier(sycl::access::fence_space::global_and_local);
```

Comment:
move to the end

Reply:
it's fine now because you still call sync after the function.
If you always call sync after function, maybe removing it here also makes sense